### PR TITLE
Detail view

### PR DESCRIPTION
**Pull request by matiasspinelli4**
## Summary ##

**Trello card**: hhttps://trello.com/c/Q1be4OGZ/16-n-6-vista-de-detalle-de-un-libro
**Trello card**: https://trello.com/c/xvIIsZ7L/17-n-61-detalle-del-libro
**Trello card**: https://trello.com/c/vYC6AuWL/18-n-62-tabla-de-comentarios
**Trello card**: https://trello.com/c/yeBRsFfy/21-n-72-extraer-componentes-en-controllers-separados
**Trello card**: https://trello.com/c/r01swp9R/22-n-73-extension-para-botones-con-borde-y-gradiente

Vista detalle de un libro

## Screenshots ##

Simulator Screen Shot - iPhone Xʀ - 2019-06-14 at 15 39 30
Simulator Screen Shot - iPhone Xʀ - 2019-06-14 at 15 39 45

## Screencast ##

.

## Known issues ##

.

### CODE REVIEW

File: MSProject/DAO/WBBookDAO.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+enum SortMethod {
+    case id
+    case title
+    case author
+    case genre
+    case year
+}
+
+class WBBookDAO: NSObject {
```
Comment (acid-wolox) [inline MSProject/DAO/WBBookDAO.swift]: I think that this whole class can be implemented in either an extension or inside a ViewModel. Maybe something that can be called like `_viewModel.getBooksSorted(by: .author)` would be really nice. But I don't know if you can do that if you really need something that sorts the books in every single view.

File: MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    
+    var delegate: DetailBookDelegate?
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+
+        backgroundColor = .clear
+    }
+    
+    var bookViewModel: WBBookViewModel? {
```
Comment (acid-wolox) [inline MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift]: The View shouldn't have an attribute referencing a ViewModel. Create a function to set all of these things and call it from the ViewController, maybe passing a `WBBookViewModel` as parameter.

File: MSProject/CustomControls/Transition.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
+//
+//  Transition.swift
+//  PushTransition
+//
+//  Created by Alejandro Zalazar on 02/05/2019.
```
[the remark this answers is not in the document]
Comment (matiasspinelli4) [reply MSProject/CustomControls/Transition.swift]: just for fun :p

File: MSProject/DAO/WBBookDAO.swift (hunk 1 above)
Comment (matiasspinelli4) [reply MSProject/DAO/WBBookDAO.swift]: at the time, I had thought this class with much more features. now it is kind a helper.

File: MSProject/Model/WBBook.swift [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    func bookStatusText() -> String {
+        switch self {
+        case .available:
+            return "AVAILABLE".localized()
+        default:
+            return "NOT_AVAILABLE".localized()
+        }
+    }
+    
+    func bookStatusAvailable() -> Bool {
```
[the remark this answers is not in the document]
Comment (DiegoQuiros94) [reply MSProject/Model/WBBook.swift]: also i would change the name to isBookAvailable, we usually name this kind of pure boolean functions/properties isSomething

File: MSProject/Networking/WBNetworkManager.swift [hunk 5]
```diff
@@ -19,26 +19,188 @@ class WBNetworkManager: NSObject {
     
     override init() {}
 
+    let userId = 5 //userID 5 ... because...
```
Comment (DiegoQuiros94) [inline MSProject/Networking/WBNetworkManager.swift]: 😆

File: MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift (hunk 2 above)
Comment (DiegoQuiros94) [reply MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift]: I think this is not wrong, but yeah, we usually use a function called setup(with: ViewModel) or something like that

File: MSProject/ViewControllers/DetailBook/ViewModel/WBBookDetailViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+//
+//  Created by Matias Spinelli on 13/06/2019.
+//  Copyright © 2019 Wolox. All rights reserved.
+//
+
+import UIKit
+
+class WBBookDetailViewModel {
+
+    private var commentsViewModels: [WBComment] = [WBComment]() {
```
Comment (DiegoQuiros94) [inline MSProject/ViewControllers/DetailBook/ViewModel/WBBookDetailViewModel.swift]: try to implement this with reactive when you get to that part of the training

**Comment by DiegoQuiros94** [comment]
Got tired of reviewing, will continue on friday :P

File: MSProject/ViewControllers/LibraryTableView/ViewModel/WBLibraryViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    
+    func selectBook(at indexPath: IndexPath) -> WBBookViewModel {
+        let book = bookViewModels[indexPath.row]
+        print("\(book.bookTitle) - \(book.bookAuthor)")
+        return book
+    }
+    
+    func sortBooks() {
+        self.libraryItems = WBBookDAO.sharedInstance.sortBooks(books: self.libraryItems, by: .id)
+        self.bookViewModels = [] //clear array
```
Comment (DiegoQuiros94) [inline MSProject/ViewControllers/LibraryTableView/ViewModel/WBLibraryViewModel.swift]: you could do `bookViewModels = libraryItems.map { WBBookViewModel($0) }`

Comment (DiegoQuiros94) [reply MSProject/ViewControllers/LibraryTableView/ViewModel/WBLibraryViewModel.swift]: also try not to use self if it is not necessary

**Comment by matiasspinelli4** [comment]
In order to keep this project simple, I decided to remove TT* classes and Transition, and use the MBProgressHUD as Instagram do (?

File: MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift [hunk 8]
```diff
@@ -36,26 +36,32 @@ class WBDetailBookHeaderView: UIView, NibLoadable {
...
+        bookAuthor.text = bookViewModel.bookAuthor
+        bookYear.text = bookViewModel.bookYear
+        bookGenre.text = bookViewModel.bookGenre
+        if bookViewModel.bookStatus == .available {
+            bookAvailable.textColor = .green
+        } else {
+            bookAvailable.textColor = .red
+        }
+        
+        if bookViewModel.bookStatus.isBookAvailable() {
```
Comment (DiegoQuiros94) [inline MSProject/ViewControllers/DetailBook/View/WBDetailBookHeaderView.swift]: is this doing the same logic as the if above? if it is you can merge them into one if

**Comment by DiegoQuiros94** [comment]
Looking good 👌 👏

**Pull request merged**